src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/LegacyDiagnosticItemSourceProvider.cs

```diff
@@ -5,40 +5,33 @@
 using System;
 using System.ComponentModel.Composition;
 using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
 using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
 using Microsoft.Internal.VisualStudio.PlatformUI;
 using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Utilities;
 
-namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer;
+
+[Export(typeof(IAttachedCollectionSourceProvider))]
+[Name(nameof(LegacyDiagnosticItemSourceProvider))]
+[Order]
+[AppliesToProject("(CSharp | VB) & !CPS")]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class LegacyDiagnosticItemSourceProvider(
+    [Import(typeof(AnalyzersCommandHandler))] IAnalyzersCommandHandler commandHandler,
+    IDiagnosticAnalyzerService diagnosticAnalyzerService) : AttachedCollectionSourceProvider<AnalyzerItem>
```

Comment:
>diagnosticAnalyzerService [](http://example.com/codeflow?start=31&length=25)

What's the rule of thumb that you use to determine whether to declare a field for these?

Reply:
if the file is 'big' :)  

if it's small (just a tiny number of methods, or like one method that then uses the PC params), then i happily use the params.  If it's large, i'm amenable to people who say they find it difficult to reason about, and i opt for fields.